File: python/python3/CSharp/Python3LexerBase.cs
using Antlr4.Runtime;
using System.Collections.Generic;
using System;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

public abstract class Python3LexerBase : Lexer {
    // A queue where extra tokens are pushed on (see the NEWLINE lexer rule).
    private LinkedList<IToken> Tokens = new LinkedList<IToken>();
    // The stack that keeps track of the indentation level.
    private Stack<int> Indents = new Stack<int>();
    // The amount of opened braces, brackets and parenthesis.
    private int Opened = 0;
    // The most recently produced token.
    private IToken LastToken = null;

    public Python3LexerBase(ICharStream input)
        : base(input)
    {
    }
    public Python3LexerBase(ICharStream input, TextWriter output, TextWriter errorOutput)
        : base(input, output, errorOutput)
    {
    }

    public override void Emit(IToken token)
    {
        base.Token = token;
        Tokens.AddLast(token);
    }

    private CommonToken CommonToken(int type, string text)
    {
        int stop = CharIndex - 1;
        int start = text.Length == 0 ? stop : stop - text.Length + 1;
        return new CommonToken(Tuple.Create((ITokenSource)this, (ICharStream)InputStream), type, DefaultTokenChannel, start, stop);
    }

    private IToken CreateDedent()
    {
        var dedent = CommonToken(Python3Parser.DEDENT, "");
        dedent.Line = LastToken.Line;
        return dedent;
    }

    public override IToken NextToken()
    {
        // Check if the end-of-file is ahead and there are still some DEDENTS expected.
        if (((ICharStream)InputStream).LA(1) == Eof && Indents.Count != 0)
        {
            // Remove any trailing EOF tokens from our buffer.
            for (var node  = Tokens.First; node != null; )
            {
                var temp = node.Next;
                if (node.Value.Type == Eof)
                {
                    Tokens.Remove(node);
                }
                node = temp;
            }
            
            // First emit an extra line break that serves as the end of the statement.
            this.Emit(CommonToken(Python3Parser.NEWLINE, "\n"));

            // Now emit as much DEDENT tokens as needed.
            while (Indents.Count != 0)
            {
                Emit(CreateDedent());
                Indents.Pop();
            }

            // Put the EOF back on the token stream.
            Emit(CommonToken(Python3Parser.Eof, "<EOF>"));
        }

        var next = base.NextToken();
        if (next.Channel == DefaultTokenChannel)
        {
            // Keep track of the last token on the default channel.
            LastToken = next;
        }

        if (Tokens.Count == 0)
        {
            return next;
        }
        else
        {
            var x = Tokens.First.Value;
            Tokens.RemoveFirst();
            return x;
        }
    }

    // Calculates the indentation of the provided spaces, taking the
    // following rules into account:
    //
    // "Tabs are replaced (from left to right) by one to eight spaces
    //  such that the total number of characters up to and including
    //  the replacement is a multiple of eight [...]"
    //
    //  -- https://docs.python.org/3.1/reference/lexical_analysis.html#indentation
    static int GetIndentationCount(string spaces)
    {
        int count = 0;
        foreach (char ch in spaces.ToCharArray())
        {
            count += ch == '\t' ? 8 - (count % 8) : 1;
        }
        return count;
    }

    public bool atStartOfInput()
    {
        return Column == 0 && Line == 1;
    }

    public void openBrace(){
        Opened++;
    }

    public void closeBrace(){
        Opened--;
    }

    public void onNewLine(){
        var newLine = (new Regex("[^\r\n\f]+")).Replace(Text, "");
        var spaces = (new Regex("[\r\n\f]+")).Replace(Text, "");

        // Strip newlines inside open clauses except if we are near EOF. We keep NEWLINEs near EOF to
        // satisfy the final newline needed by the single_put rule used by the REPL.
        int next = ((ICharStream)InputStream).LA(1);
        int nextnext = ((ICharStream)InputStream).LA(2);
        if (Opened > 0 || (nextnext != -1 && (next == '\r' || next == '\n' || next == '\f' || next == '#')))
        {
            // If we're inside a list or on a blank line, ignore all indents, 
            // dedents and line breaks.
            Skip();
        }
        else
        {
            Emit(CommonToken(Python3Lexer.NEWLINE, newLine));
            int indent = GetIndentationCount(spaces);
            int previous = Indents.Count == 0 ? 0 : Indents.Peek();
            if (indent == previous)
            {
                // skip indents of the same size as the present indent-size
                Skip();
            }
            else if (indent > previous) {
                Indents.Push(indent);
                Emit(CommonToken(Python3Parser.INDENT, spaces));
            }
            else {
                // Possibly emit more than 1 DEDENT token.
                while(Indents.Count != 0 && Indents.Peek() > indent)
                {
                    this.Emit(CreateDedent());
                    Indents.Pop();
                }
            }
        }
    }

    public override void Reset()
    {
        Tokens = new LinkedList<IToken>();
        Indents = new Stack<int>();
        Opened = 0;
        LastToken = null;
        base.Reset();
    }
}

File: python/python3/TypeScript/Python3LexerBase.ts
import { CommonToken } from "antlr4ts/CommonToken";
import { Lexer } from "antlr4ts/Lexer";
import { Python3Parser } from './Python3Parser';

export default abstract class Python3LexerBase extends Lexer {
    tokens: any[];
    indents: any[];
    opened: number;

    constructor(input) {
        super(input);
        this.tokens = [];
        this.indents = [];
        this.opened = 0;
    }

    reset() {
        // A queue where extra tokens are pushed on (see the NEWLINE lexer rule).
        this.tokens = [];

        // The stack that keeps track of the indentation level.
        this.indents = [];

        // The amount of opened braces, brackets and parenthesis.
        this.opened = 0;

        super.reset();
    }

    emitToken(token) {
        this._token = token;
        this.tokens.push(token);
    }

    nextToken() {
        // Check if the end-of-file is ahead and there are still some DEDENTS expected.
        if (this._input.LA(1) === Python3Parser.EOF && this.indents.length) {
            // Remove any trailing EOF tokens from our buffer.
            this.tokens = this.tokens.filter(function (val) {
                return val.type !== Python3Parser.EOF;
            });
            // First emit an extra line break that serves as the end of the statement.
            this.emitToken(this.commonToken(Python3Parser.NEWLINE, "\n"));
            // Now emit as much DEDENT tokens as needed.
            while (this.indents.length) {
                this.emitToken(this.createDedent());
                this.indents.pop();
            }
            // Put the EOF back on the token stream.
            this.emitToken(this.commonToken(Python3Parser.EOF, "<EOF>"));
        }
        let next = super.nextToken();
        return this.tokens.length ? this.tokens.shift() : next;
    }

    createDedent() {
        return this.commonToken(Python3Parser.DEDENT, "");
    }

    getCharIndex() {
		return this._input.index;
	}

    commonToken(type, text) {
        let stop = this.getCharIndex() - 1;
        let start = text.length ? stop - text.length + 1 : stop;
        return new CommonToken(type, text, this._tokenFactorySourcePair, Lexer.DEFAULT_TOKEN_CHANNEL, start, stop);
    }

    getIndentationCount(whitespace) {
        let count = 0;
        for (let i = 0; i < whitespace.length; i++) {
            if (whitespace[i] === '\t') {
                count += 8 - count % 8;
            } else {
                count++;
            }
        }
        return count;
    }

    atStartOfInput() {
        return this.getCharIndex() === 0;
    }

    openBrace() {
        this.opened++;
    }

    closeBrace() {
        this.opened--;
    }

    onNewLine() {
        let newLine = this.text.replace(/[^\r\n]+/g, '');
        let spaces = this.text.replace(/[\r\n]+/g, '');

        // Strip newlines inside open clauses except if we are near EOF. We keep NEWLINEs near EOF to
        // satisfy the final newline needed by the single_put rule used by the REPL.
        let next = this._input.LA(1);
        let nextnext = this._input.LA(2);
        if (this.opened > 0 || (nextnext != -1 /* EOF */ && (next === 13 /* '\r' */ || next === 10 /* '\n' */ || next === 35 /* '#' */))) {
            // If we're inside a list or on a blank line, ignore all indents,
            // dedents and line breaks.
            this.skip();
        } else {
            this.emitToken(this.commonToken(Python3Parser.NEWLINE, newLine));

            let indent = this.getIndentationCount(spaces);
            let previous = this.indents.length ? this.indents[this.indents.length - 1] : 0;

            if (indent === previous) {
                // skip indents of the same size as the present indent-size
                this.skip();
            } else if (indent > previous) {
                this.indents.push(indent);
                this.emitToken(this.commonToken(Python3Parser.INDENT, spaces));
            } else {
                // Possibly emit more than 1 DEDENT token.
                while (this.indents.length && this.indents[this.indents.length - 1] > indent) {
                    this.emitToken(this.createDedent());
                    this.indents.pop();
                }
            }
        }
    }
}

